src/utils/taint.js

```diff
@@ -0,0 +1,398 @@
+      !(val instanceof NativeFunctionValue),
+      "all native function values should have already been created outside this pure function"
+    );
+
+    let environment = val.$Environment.parent;
+    while (environment) {
+      let record = environment.environmentRecord;
+      if (record instanceof ObjectEnvironmentRecord) {
+        this.visitValue(record.object);
+        continue;
+      }
+
+      invariant(
+        !(record instanceof GlobalEnvironmentRecord),
+        "we should never reach the global scope because it is never impure in a pure function."
+      );
+      invariant(record instanceof DeclarativeEnvironmentRecord);
+
+      this.visitDeclarativeEnvironmentRecordBinding(record);
+
+      if (record instanceof FunctionEnvironmentRecord) {
+        // If this is a function environment, we visit the function object which if it is
+        // impure will comeback here to visit its parent environment.
+        let fn = record.$FunctionObject;
+        this.visitValue(fn);
+        break;
+      }
+      environment = environment.parent;
+    }
+  }
+
+  visitValueObject(val: ObjectValue): void {
+    let kind = val.getKind();
+    this.visitObjectProperties(val, kind);
+
+    switch (kind) {
+      case "RegExp":
+      case "Number":
+      case "String":
+      case "Boolean":
+      case "ReactElement":
+      case "ArrayBuffer":
+        return;
+      case "Date":
+        let dateValue = val.$DateValue;
+        invariant(dateValue !== undefined);
+        this.visitValue(dateValue);
+        return;
+      case "Float32Array":
+      case "Float64Array":
+      case "Int8Array":
+      case "Int16Array":
+      case "Int32Array":
+      case "Uint8Array":
+      case "Uint16Array":
+      case "Uint32Array":
+      case "Uint8ClampedArray":
+      case "DataView":
+        let buf = val.$ViewedArrayBuffer;
+        invariant(buf !== undefined);
+        this.visitValue(buf);
+        return;
+      case "Map":
+      case "WeakMap":
+        this.visitValueMap(val);
+        return;
+      case "Set":
+      case "WeakSet":
+        this.visitValueSet(val);
+        return;
+      default:
+        invariant(kind === "Object", `Object of kind ${kind} is not supported in calls to abstract functions.`);
+        invariant(
+          this.$ParameterMap === undefined,
+          `Arguments object is not supported in calls to abstract functions.`
+        );
+        return;
+    }
+  }
+
+  visitValueSymbol(val: SymbolValue): void {
+    if (val.$Description) this.visitValue(val.$Description);
+  }
+
+  visitValueProxy(val: ProxyValue): void {
+    this.visitValue(val.$ProxyTarget);
+    this.visitValue(val.$ProxyHandler);
+  }
+
+  visitAbstractValue(val: AbstractValue): void {
+    for (let i = 0, n = val.args.length; i < n; i++) {
+      this.visitValue(val.args[i]);
+    }
+  }
+
+  visitValue(val: Value): void {
+    if (val instanceof AbstractValue) {
+      if (this._mark(val)) this.visitAbstractValue(val);
```

Comment:
_mark will never return true for an abstract value.

Reply:
Yea this is surfacing a bug in `_mark` which I'll fix by only bailing out if it's an `ObjectValue` and newly created.